### PR TITLE
Add CI typechecking

### PR DESCRIPTION
**Pull request by abejgonzalez**
Fixes CI type checking errors and adds a CI type checking step to CI. Builds off #1316 

### Contributor Checklist
- [x] Is this PR's title suitable for inclusion in the changelog and have you added a `changelog:<topic>` label?
- [x] Did you add Scaladoc/docstring/doxygen to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous prints/debugging code?
- [x] Did you state the UI / API impact?
- [x] Did you specify the Verilog / AGFI compatibility impact?

- [x] If applicable, did you regenerate and publicly share default AGFIs?

- [x] If applicable, did you apply the `ci:fpga-deploy` label?

- [x] If applicable, did you apply the `Please Backport` label?

### Reviewer Checklist (only modified by reviewer)
Note: to run CI on PRs from forks, comment `@Mergifyio copy main` and manage the change from the new PR.
- [ ] Is the title suitable for inclusion in the changelog and does the PR have a `changelog:<topic>` label?
- [ ] Did you mark the proper release milestone?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?

### CODE REVIEW

File: .github/workflows/firesim-run-tests.yml
Comment (abejgonzalez) [inline .github/workflows/firesim-run-tests.yml:183]:
```suggestion
        TERM: xterm-256-color
```

File: .github/scripts/build-default-workloads.py
Comment (t14916) [inline .github/scripts/build-default-workloads.py:3]: So this ignores all the typing information coming out of functions in fabric.api right? I've looked into this in the past,  and don't remember seeing any way to get the types out of fabric.api, but maybe it's worth looking into again?

File: .github/scripts/platform_lib.py
Comment (t14916) [inline .github/scripts/platform_lib.py:256]: So the reason this is done is to avoid some of the typing issues with the boto3 client?

File: .github/scripts/github_common.py
Comment (t14916) [inline .github/scripts/github_common.py:20]: Is it possible to be more specific than Any? since we're pulling down a json from http, is it possible to get something that isn't either `int` or `string`? This is safer for sure.

File: .github/workflows/firesim-run-tests.yml
Comment (t14916) [inline .github/workflows/firesim-run-tests.yml:178]: Just a thought, but what if we made the ci typecheck the first script run in the CI pipeline? This might be a bit too strict, but the idea would be to check earlier on and maybe make this a dependency that forces us to fix typing before testing CI. Just a suggestion, this might be to anal.

File: .github/workflows/firesim-run-tests.yml
Comment (abejgonzalez) [inline .github/workflows/firesim-run-tests.yml:187]:
```suggestion
        run: .github/scripts/run-ci-python-typechecking.py
```

**Comment by t14916** [comment]
just for personal reference, did you make these changes by hand or did you use some tool to fix some of the mypy errors (expanding the wildcards, adding the type ignore comments,etc.)

File: .github/workflows/firesim-run-tests.yml
Comment (abejgonzalez) [reply .github/workflows/firesim-run-tests.yml:178]: I agree this should be done. However, I think the best way to do this is to setup a way to initialize the FireSim repo w/ GH-A managed runners and move this job there so that even spawning the manager is blocked by a typing check. I don't have time to do this, so I'll leave this to you or just future work.

Comment (t14916) [reply .github/workflows/firesim-run-tests.yml:178]: Yeah that's what I was suggesting, I'll put it in my backlog

File: .github/scripts/platform_lib.py
Comment (abejgonzalez) [reply .github/scripts/platform_lib.py:256]: In the `cull_instances` file you explicitly call the `terminate_azure_vms` which is a hidden method in the platform library. So I added this new method so that we avoid piping platform-specific methods out from the platform lib.

File: .github/scripts/build-default-workloads.py
Comment (abejgonzalez) [reply .github/scripts/build-default-workloads.py:3]: Correct. Fabric doesn't have any types. I also since we use a hacked version of Fabric (fab-classic) I don't expect it to add types anytime soon. We use this "hack" across the codebase in the interim.

File: .github/scripts/github_common.py
Comment (abejgonzalez) [reply .github/scripts/github_common.py:20]: Technically the dict can have values of lists, strs, dicts, bools, floats, etc so its a bit more complicated. Looking around quickly there is a way to specify a JSON type manually (https://github.com/python/typing/issues/182) but I'm not sure where to put the alias definition (I think it would be good to avoid duplicating this in places). I figured this was sufficient since we do this in other places but we can make this an issue for future work.

**Comment by abejgonzalez** [comment]
> just for personal reference, did you make these changes by hand or did you use some tool to fix some of the mypy errors (expanding the wildcards, adding the type ignore comments,etc.)

I did this by hand.

**Review by t14916 — APPROVED** [review]
lgtm

**Pull request merged**
